engines/prince/prince.cpp

```diff
@@ -682,11 +829,175 @@ void PrinceEngine::showTexts(Graphics::Surface *screen) {
+			}
+
+			// Credits in other languages may have some Polish characters
+			inTable = checkConversionTable(b, i, convertedBytes, polishEncodingTable);
+		}
+
+		if (!inTable) {
+			convertedBytes[i] = *b;
+			i++;
+		}
+	}
+
+	convertedBytes[i] = 0;
+
+	Common::U32String result((char *)convertedBytes);
+	delete[] convertedBytes;
+
+	return result;
+#endif
+}
+
+bool PrinceEngine::checkConversionTable(const byte *character, int &index, byte *convertedBytes, const int *table) const {
+	for (int i = 0; table[i]; i += 2) {
+		if (*character == table[i]) {
+			convertedBytes[index] = (table[i + 1] >> 8) & 0xff;
+			convertedBytes[index + 1] = table[i + 1] & 0xff;
+			index += 2;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+void PrinceEngine::setTTSVoice(uint32 textColor) const {
+#ifdef USE_TTS
+	Common::TextToSpeechManager *ttsMan = g_system->getTextToSpeechManager();
+	if (ttsMan != nullptr && (ConfMan.getBool("tts_enabled_speech") || ConfMan.getBool("tts_enabled_objects") || ConfMan.getBool("tts_enabled_missing_voice"))) {
+		int id = 0;
+
+		for (int i = 0; i < kNumberOfCharacterDatas; ++i) {
+			// In many cases, characters can be differentiated by just the text color, but sometimes
+			// there may be different characters with the same text colors in different locations, and rarely
+			// different characters with the same text colors in the same location. Using the location number and/or
+			// mob index differentiates characters in these cases
+			if (characterVoiceDatas[i].textColor == textColor && 
+				(characterVoiceDatas[i].locationNumber == 0 || characterVoiceDatas[i].locationNumber == _locationNr) &&
+				(characterVoiceDatas[i].mobIndex == -1 || characterVoiceDatas[i].mobIndex == _dialogMob)) {
+				id = i;
+				break;
+			}
+		}
+
+		Common::Array<int> voices;
+		int pitch = 0;
+		Common::TTSVoice::Gender gender;
+
+		if (characterVoiceDatas[id].male) {
+			voices = ttsMan->getVoiceIndicesByGender(Common::TTSVoice::MALE);
+			gender = Common::TTSVoice::MALE;
+		} else {
+			voices = ttsMan->getVoiceIndicesByGender(Common::TTSVoice::FEMALE);
+			gender = Common::TTSVoice::FEMALE;
+		}
+
+		// If no voice is available for the necessary gender, set the voice to default
+		if (voices.empty()) {
+			ttsMan->setVoice(0);
+		} else {
+			int voiceIndex = characterVoiceDatas[id].voiceID % voices.size();
+			ttsMan->setVoice(voices[voiceIndex]);
+		}
+
+		// If no voices are available for this gender, alter the pitch to mimic a voice
+		// of the other gender
+		if (ttsMan->getVoice().getGender() != gender) {
+			if (gender == Common::TTSVoice::MALE) {
+				pitch -= 50;
+			} else {
+				pitch += 50;
+			}
+		}
+
+		ttsMan->setPitch(pitch);
+	}
+#endif
+}
+
+void PrinceEngine::stopTextToSpeech() const {
+	Common::TextToSpeechManager *ttsMan = g_system->getTextToSpeechManager();
+	if (ttsMan != nullptr && (ConfMan.getBool("tts_enabled_objects") || ConfMan.getBool("tts_enabled_speech") || ConfMan.getBool("tts_enabled_missing_voice")) &&
```

Comment:
I am not sure about this one... What if the intro plays, you go into the scummvm options menu to disable the tts features - but the voice is still speaking (and the game continues after you've left the scummvm options menu)